pgx/src/spi.rs

```diff
@@ -378,8 +378,176 @@ impl SpiClient {
+                src.as_ptr(),
+                nargs as i32,
+                argtypes.as_mut_ptr(),
+                datums.as_mut_ptr(),
+                nulls.as_ptr(),
+                false,
+                0,
+            )
+        };
+        SpiCursor { ptr }
+    }
+
+    /// find a cursor in transaction by name.
+    /// Name should usually be retrieved via [`SpiCursor::into_name`]
+    ///
+    /// See [`SpiCursor`] docs for usage details.
+    pub fn find_cursor(
+        &mut self,
+        name: String,
+    ) -> SpiCursor {
+        use pgx_pg_sys::AsPgCStr;
+
+        let ptr = unsafe {
+            pg_sys::SPI_cursor_find(name.as_pg_cstr())
+        };
+        if ptr.is_null() {
+            error!("cursor named \"{}\" not found", name);
+        }
+        SpiCursor { ptr }
+    }
+}
+
+/// Represents a Postgres cursor (internally, a portal), allowing to retrieve result rows a few
+/// at a time. Moreover, a cursor can be left open within a transaction, and accessed in
+/// multiple independent Spi sessions within the transaction.
+///
+/// A cursor can be created via [`SpiClient::open_cursor()`] from a query; a cursor left open (i.e.
+/// not [`Self::close()`]d) can be turned into a cursor name via [`Self::into_name()`] and later
+/// retrieved by name (in the same transaction) via [`SpiClient::find_cursor()`].
+///
+/// # Important notes about memory usage
+/// Result sets returned by [`SpiCursor::fetch()`] will not be freed until the current
+/// Spi session is complete; this is a Pgx limitation that might get lifted in the future.
+///
+/// In the meantime, if you're using cursors to limit memory usage, make sure to use
+/// multiple separate Spi sessions, retrieving the cursor by name.
+///
+/// # Examples
+/// ## Simple cursor
+/// ```rust,no_run
+/// use pgx::Spi;
+/// Spi::connect(|mut client| {
+///     let mut cursor = client.open_cursor("SELECT * FROM generate_series(1, 5)", None);
+///     assert_eq!(Some(1u32), cursor.fetch(1).get_one());
+///     assert_eq!(Some(2u32), cursor.fetch(2).get_one());
+///     assert_eq!(Some(3u32), cursor.fetch(3).get_one());
+///     Ok(None::<()>)
+///     // <--- all three result sets get freed only at this point
+/// });
+/// ```
+///
+/// ## Cursor by name
+/// ```rust,no_run
+/// use pgx::Spi;
+/// let cursor_name = Spi::connect(|mut client| {
+///     let mut cursor = client.open_cursor("SELECT * FROM generate_series(1, 5)", None);
+///     assert_eq!(Some(1u32), cursor.fetch(1).get_one());
+///     Ok(Some(cursor.into_name()))
+///     // <--- first result set gets freed at this point
+/// }).unwrap();
+/// Spi::connect(|mut client| {
+///     let mut cursor = client.find_cursor(cursor_name);
+///     assert_eq!(Some(2u32), cursor.fetch(1).get_one());
+///     Ok(None::<()>)
+///     // <--- second result set gets freed at this point
+/// });
+/// ```
+pub struct SpiCursor {
+    ptr: pg_sys::Portal,
 }
 
+impl SpiCursor {
+    /// fetch some rows from a cursor
```

Comment:
```suggestion
    /// Fetch `count` rows from the cursor
```

I'd like to also specify what the behavior is if the count is more than the number of rows available.

Reply:
👍 Done